rrdd/rrdd_server.ml

```diff
@@ -274,31 +274,57 @@ let forget_host_ds (ds_name : string) : unit =
     )
 
 let query_possible_dss rrdi =
+  (* We have data sources coming from different places, so we want the union of these;
+     this happens to be the union of the rrdi.rrd (live) and rrdi.rrd.rdd_dss (other).
+     This is straighforward, but those two sets of rrds have different types,
+     hence we need to coerce them into a common type (this is why we have
+     'description not available', and 'units unknown' etc.).
+     Deciding whether a ds is enabled is also not obvious. If we have a 'live'
+     ds, then it is enabled if it exists in the set rrdi.rrd. If we have an 'other'
+     ds, then it is enabled if it is also an enabled 'live' ds, otherwise it is
+     disabled.
+   *)
   let module SMap = Map.Make(String) in
-  let enabled_sources = rrdi.dss
-  |> List.to_seq
-  |> Seq.map (fun ds -> ds.Ds.ds_name, ds)
-  |> SMap.of_seq
+  let open Ds in
+  let open Data_source in
+
+  let live_sources = rrdi.dss in
+  let other_sources = rrdi.rrd.rrd_dss in
+
+  (* we identify dss by their names. assume names which differ only by whitespace or case are actually the same *)
+  let normalize str = str |> Astring.String.trim |> Astring.String.Ascii.lowercase in
+  let map = (* we use the map variable to accumulate all dss *)
+    let is_live_ds_enabled ds = List.mem (normalize ds.ds_name) (Rrd.ds_names rrdi.rrd |> List.map normalize) in
+    live_sources
+    |> List.to_seq
+    |> Seq.map (fun ds -> normalize ds.ds_name,
+      { name = ds.ds_name
+      ; description = ds.ds_description
+      ; enabled = is_live_ds_enabled ds
+      ; standard = ds.ds_default
+      ; min= ds.ds_min
+      ; max = ds.ds_max
+      ; units = ds.ds_units
+      })
+    |> SMap.of_seq
+  in
+  let map = Array.fold_left (fun map next_ds ->
+    let name = normalize next_ds.Rrd.ds_name in
+    if SMap.mem name map then map
+    else
+      let new_elt = {
+          name = next_ds.ds_name
+        ; description = "description not available"
+        ; enabled = false
+        ; standard = false
+        ; units = "unknown"
+        ; min = next_ds.ds_min
+        ; max = next_ds.ds_max
+        }
+      in SMap.add name new_elt map
+      ) map other_sources
```

Comment:
I would prefer an approach that make it obvious what is happening in this block. i.e. we're adding 'disabled archival' sources.
Something like:
```
let archival_disabled_names_to_sources = List.filter_map (... if disabled then Some (...) else None) archival_sources in
let all_names_to_sources = SMap.union (fun ... -> throw warning and pick from first map) map archival_disabled_names_to_sources in
all_names_to_sources |> Smap.bindings |> List.map snd
```